src/ts_catalog/dimension_partition.c

```diff
@@ -0,0 +1,432 @@
+	ts_scan_iterator_close(&it);
+}
+
+/*
+ * Delete all dimension partitions for a given dimension.
+ */
+void
+ts_dimension_partition_info_delete(int dimension_id)
+{
+	dimension_partition_info_delete(dimension_id, SCANNER_F_NOFLAGS, RowExclusiveLock);
+}
+
+DimensionPartitionInfo *
+ts_dimension_partition_info_recreate(int32 dimension_id, unsigned int num_partitions,
+									 List *data_nodes, int replication_factor)
+{
+	int64 partition_size = DIMENSION_SLICE_CLOSED_MAX / ((int64) num_partitions);
+	int64 range_start = DIMENSION_SLICE_MINVALUE;
+	Catalog *catalog = ts_catalog_get();
+	Oid relid = catalog_get_table_id(catalog, DIMENSION_PARTITION);
+	DimensionPartitionInfo *dpi;
+	RBTree *tree = NULL;
+	Relation rel;
+	unsigned int i;
+
+	Assert(num_partitions > 0);
+	Assert(data_nodes == NIL || replication_factor > 0);
+
+	/* Delete all partitions for the dimension */
+	dimension_partition_info_delete(dimension_id, SCANNER_F_KEEPLOCK, RowExclusiveLock);
+
+	/* Lock already held */
+	rel = table_open(relid, NoLock);
+
+	tree = rbt_create(sizeof(DimensionPartition),
+					  dimpart_cmp,
+					  dimpart_combine,
+					  dimpart_alloc,
+					  dimpart_free,
+					  NULL);
+
+	for (i = 0; i < num_partitions; i++)
+	{
+		bool is_new = false;
+		int64 range_end =
+			(i == (num_partitions - 1)) ? DIMENSION_SLICE_CLOSED_MAX : range_start + partition_size;
+		DimensionPartition dp = {
+			.dimension_id = dimension_id,
+			.range_start = range_start,
+			.range_end = range_end,
+			.data_nodes = get_replica_nodes(data_nodes, i, replication_factor),
+		};
+		CatalogSecurityContext sec_ctx;
+		HeapTuple tuple;
+
+		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+		tuple = create_dimension_partition_tuple(rel, &dp);
+		ts_catalog_insert_only(rel, tuple);
+		ts_catalog_restore_user(&sec_ctx);
+		heap_freetuple(tuple);
+		rbt_insert(tree, &dp.rbtnode, &is_new);
+
+		/* Hash values for space partitions are in range 0 to INT32_MAX, so
+		 * the first partition covers 0 to partition size (although the start
+		 * value is -INF) */
+		if (range_start == DIMENSION_SLICE_MINVALUE)
+			range_start = 0;
+
+		range_start += partition_size;
+	}
+
+	table_close(rel, RowExclusiveLock);
+
+	/* Make changes visible */
+	CommandCounterIncrement();
+
+	dpi = palloc(sizeof(DimensionPartitionInfo));
+	dpi->partitions = tree;
+	dpi->num_partitions = num_partitions;
+
+	return dpi;
+}
+
+/*
+ * Manually update the dimension partition state for a hypertable.
+ *
+ * Used mostly when updating from a TimescaleDB version that didn't have this
+ * state previously.
+ */
+TS_FUNCTION_INFO_V1(ts_dimension_partition_update);
+
+Datum
+ts_dimension_partition_update(PG_FUNCTION_ARGS)
+{
+	Oid hypertable_relid = PG_GETARG_OID(0);
+	Cache *hcache;
+	const Hypertable *ht =
+		ts_hypertable_cache_get_cache_and_entry(hypertable_relid, CACHE_FLAG_NONE, &hcache);
+	ts_hypertable_update_dimension_partitions(ht);
+	ts_cache_release(hcache);
+	PG_RETURN_VOID();
```

Comment:
In general, this file has a lot of warnings about less code coverage. We can try to increase tests which cover atleast some of the warnings

Reply:
Switched to array+binary search and this should improve test coverage as there will be fever unused rb-tree functions.